crates/bevy_ecs/src/system/query.rs

```diff
@@ -363,7 +363,7 @@ impl<'w, 's, Q: WorldQueryData, F: WorldQueryFilter> Query<'w, 's, Q, F> {
     /// This will create a query that could violate memory safety rules. Make sure that this is only
     /// called in ways that ensure the queries have unique mutable access.
```

Comment:
```suggestion
    /// `world` must have permission to access any component data specified in `state`'s [`archetype_component_access`].
    ///
    /// [`archetype_component_access`]: QueryState::archetype_component_access
    ///
    /// # Panics
    ///
    /// If `world` does not match the one used to create `state`.
```

I believe this is the most correct way you can phrase the safety requirement. Though we'd also need to add `QueryState::archetype_component_access` (it's weird that function doesn't exist already).

The `# Panics` should probably go above the `# Safety` section, although GitHub's UI won't let me format the suggestion that way.